# Importing Libraries

In [1]:
import pandas as pd
import numpy as np
import os

Importing Data

In [2]:
# Folder path
path = r'/Users/michaelyeboah/Instacart Basket Analysis'

In [3]:
path

'/Users/michaelyeboah/Instacart Basket Analysis'

In [4]:
# Importing 'orders_products_merged_updated.pkl'

df_orders_products_merged = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_updated.pkl'))

In [5]:
df_orders_products_merged.head()

order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329        1    prior             1                   2   
1   2398795        1    prior             2                   3   
2    473747        1    prior             3                   3   
3   2254736        1    prior             4                   4   
4    431534        1    prior             5                   4   

   order_hour_of_day  days_since_last_order  product_id  add_to_cart_order  \
0                  8                    NaN         196                  1   
1                  7                   15.0         196                  1   
2                 12                   21.0         196                  1   
3                  7                   29.0         196                  1   
4                 15                   28.0         196                  1   

   reordered  Unnamed: 0_y product_name  aisle_id  department_id  prices  \
0          0           195         Soda        77              7     9.0   
1          1           195         Soda        77              7     9.0   
2          1           195         Soda        77              7     9.0   
3          1           195         Soda        77              7     9.0   
4          1           195         Soda        77              7     9.0   

  _merge     busiest_day     busiest_days busiest_period_of_day  
0   both  Regularly busy   Regularly busy        Average orders  
1   both  Regularly busy  Least busy days        Average orders  
2   both  Regularly busy  Least busy days           Most orders  
3   both      Least busy  Least busy days        Average orders  
4   both      Least busy  Least busy days           Most orders

In [6]:
df_orders_products_merged.shape

(32404859, 19)

In [7]:
# Creating subset

df = df_orders_products_merged[:1000000]

In [8]:
df.head(10)

order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329        1    prior             1                   2   
1   2398795        1    prior             2                   3   
2    473747        1    prior             3                   3   
3   2254736        1    prior             4                   4   
4    431534        1    prior             5                   4   
5   3367565        1    prior             6                   2   
6    550135        1    prior             7                   1   
7   3108588        1    prior             8                   1   
8   2295261        1    prior             9                   1   
9   2550362        1    prior            10                   4   

   order_hour_of_day  days_since_last_order  product_id  add_to_cart_order  \
0                  8                    NaN         196                  1   
1                  7                   15.0         196                  1   
2                 12                   21.0         196                  1   
3                  7                   29.0         196                  1   
4                 15                   28.0         196                  1   
5                  7                   19.0         196                  1   
6                  9                   20.0         196                  1   
7                 14                   14.0         196                  2   
8                 16                    0.0         196                  4   
9                  8                   30.0         196                  1   

   reordered  Unnamed: 0_y product_name  aisle_id  department_id  prices  \
0          0           195         Soda        77              7     9.0   
1          1           195         Soda        77              7     9.0   
2          1           195         Soda        77              7     9.0   
3          1           195         Soda        77              7     9.0   
4          1           195         Soda        77              7     9.0   
5          1           195         Soda        77              7     9.0   
6          1           195         Soda        77              7     9.0   
7          1           195         Soda        77              7     9.0   
8          1           195         Soda        77              7     9.0   
9          1           195         Soda        77              7     9.0   

  _merge     busiest_day     busiest_days busiest_period_of_day  
0   both  Regularly busy   Regularly busy        Average orders  
1   both  Regularly busy  Least busy days        Average orders  
2   both  Regularly busy  Least busy days           Most orders  
3   both      Least busy  Least busy days        Average orders  
4   both      Least busy  Least busy days           Most orders  
5   both  Regularly busy   Regularly busy        Average orders  
6   both  Regularly busy     Busiest days           Most orders  
7   both  Regularly busy     Busiest days           Most orders  
8   both  Regularly busy     Busiest days           Most orders  
9   both      Least busy  Least busy days        Average orders

# Grouping Data with Pandas

In [9]:
df.groupby('product_name')

In [ ]:
# To recap, you should always use the groupby() function as part of a series of steps, namely, the following:

# 1. Split the data into groups based on some criteria.
# 2. Apply a function to each group separately.
# 3. Combine the results into a dataframe or alternative data structure or create a new column in the current dataframe.

# So far, you’ve only completed the first step—splitting the data into groups (with the group being the “product_name” column)

Aggregating Data with agg()

In [ ]:
#  If you were to calculate the mean of the “order_number” column grouped by the “department_id” column,
# you could quickly compare the average number of orders across each Instacart department.
# This will involve two of the steps in the three-step process introduced above:

# 1. Split the data into groups based on “department_id.”
# 2. Apply the agg() function to each group to obtain the mean values for the “order_number” column.

Single Aggregation

In [10]:
df.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
4                18.825780
7                17.472355
13               17.993423
14               19.246334
16               19.463012
17               11.294069
19               19.305237
20               17.599636

In [11]:
# Different way of finding mean of 'order_number'

df.groupby('department_id')['order_number'].mean()

department_id
4     18.825780
7     17.472355
13    17.993423
14    19.246334
16    19.463012
17    11.294069
19    19.305237
20    17.599636
Name: order_number, dtype: float64

Multiple Aggregations

In [12]:
df.groupby('department_id').agg({'order_number': ['mean', 'min', 'max']})

order_number        
                      mean min max
department_id                     
4                18.825780   1  99
7                17.472355   1  99
13               17.993423   1  99
14               19.246334   1  99
16               19.463012   1  99
17               11.294069   1  98
19               19.305237   1  99
20               17.599636   1  99

Aggregating Data with transform()

In [37]:
# Now, let’s map this task onto the three-step process introduced earlier:

# 1. Split the data into groups based on the “user_id” column.
# 2. Apply the transform() function on the “order_number” column to generate the maximum orders for each user.
# 3. Create a new column, “max_order,” into which you’ll place the results of your aggregation.

In [13]:
df_orders_products_merged['max_order'] = df_orders_products_merged.groupby('user_id')['order_number'].transform(np.max)
# df_orders_products_merged['max_order'] = df_orders_products_merged.groupby(['user_id'])['order_number'].transform(np.max)

# Let’s run through each piece of the code in turn:
# First, a new column called “max_order” is created, which will be what stores the maximum order number for each user (step 3).
# Then, the df_orders_products_merged dataframe is grouped by the “user_id” column (step 1).
# And finally, the transform() function is applied on the “order_number” column with the np.max argument (step 2).

In [14]:
df_orders_products_merged.head(25)

order_id  user_id eval_set  order_number  orders_day_of_week  \
0    2539329        1    prior             1                   2   
1    2398795        1    prior             2                   3   
2     473747        1    prior             3                   3   
3    2254736        1    prior             4                   4   
4     431534        1    prior             5                   4   
5    3367565        1    prior             6                   2   
6     550135        1    prior             7                   1   
7    3108588        1    prior             8                   1   
8    2295261        1    prior             9                   1   
9    2550362        1    prior            10                   4   
10   2968173       15    prior            15                   1   
11   1870022       15    prior            17                   2   
12   1911383       15    prior            18                   2   
13   2715276       15    prior            21                   1   
14    487368       15    prior            22                   1   
15   2293453       19    prior             2                   5   
16   1973799       19    prior             5                   6   
17    532817       19    prior             7                   4   
18   1573906       21    prior            10                   3   
19   1593000       31    prior            10                   3   
20   2231262       31    prior            17                   3   
21   2580647       43    prior             6                   4   
22   2187180       43    prior             9                   4   
23   2497897       52    prior             1                   1   
24   1318871       52    prior             2                   1   

    order_hour_of_day  days_since_last_order  product_id  add_to_cart_order  \
0                   8                    NaN         196                  1   
1                   7                   15.0         196                  1   
2                  12                   21.0         196                  1   
3                   7                   29.0         196                  1   
4                  15                   28.0         196                  1   
5                   7                   19.0         196                  1   
6                   9                   20.0         196                  1   
7                  14                   14.0         196                  2   
8                  16                    0.0         196                  4   
9                   8                   30.0         196                  1   
10                  9                    7.0         196                  2   
11                 16                    8.0         196                  6   
12                 11                    7.0         196                  1   
13                  9                    7.0         196                  1   
14                 10                   14.0         196                  1   
15                 14                    6.0         196                  3   
16                 12                    8.0         196                 15   
17                 17                    6.0         196                  1   
18                 10                    6.0         196                  2   
19                  8                    7.0         196                 17   
20                 11                    8.0         196                 14   
21                 16                    4.0         196                  9   
22                 12                    3.0         196                  1   
23                  9                    NaN         196                  4   
24                 10                    7.0         196                  5   

    reordered  Unnamed: 0_y product_name  aisle_id  department_id  prices  \
0           0           195         Soda        77              7     9.0   
1           1           19

Deriving Columns with loc()

In [18]:
# To create your flag, you’ll need some criteria. You can use the following:

# 1. If the maximum orders the user has made is over 40, then the customer will be labeled a “Loyal customer.”
# 2. If the maximum orders the user has made is over 10 but less than or equal to 40, then the customer will be labeled a “Regular customer.”
# 3. If the maximum orders the user has made is less than or equal to 10, then the customer will be labeled a “New customer.”

In [15]:
df_orders_products_merged.loc[df_orders_products_merged['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [16]:
df_orders_products_merged.loc[(df_orders_products_merged['max_order'] <= 40) & (df_orders_products_merged['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [17]:
df_orders_products_merged.loc[df_orders_products_merged['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [18]:
df_orders_products_merged['loyalty_flag'].value_counts(dropna = False)

Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: loyalty_flag, dtype: int64

In [19]:
df_orders_products_merged.head()

order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329        1    prior             1                   2   
1   2398795        1    prior             2                   3   
2    473747        1    prior             3                   3   
3   2254736        1    prior             4                   4   
4    431534        1    prior             5                   4   

   order_hour_of_day  days_since_last_order  product_id  add_to_cart_order  \
0                  8                    NaN         196                  1   
1                  7                   15.0         196                  1   
2                 12                   21.0         196                  1   
3                  7                   29.0         196                  1   
4                 15                   28.0         196                  1   

   reordered  ...  product_name aisle_id  department_id  prices  _merge  \
0          0  ...          Soda       77              7     9.0    both   
1          1  ...          Soda       77              7     9.0    both   
2          1  ...          Soda       77              7     9.0    both   
3          1  ...          Soda       77              7     9.0    both   
4          1  ...          Soda       77              7     9.0    both   

      busiest_day     busiest_days busiest_period_of_day max_order  \
0  Regularly busy   Regularly busy        Average orders        10   
1  Regularly busy  Least busy days        Average orders        10   
2  Regularly busy  Least busy days           Most orders        10   
3      Least busy  Least busy days        Average orders        10   
4      Least busy  Least busy days           Most orders        10   

   loyalty_flag  
0  New customer  
1  New customer  
2  New customer  
3  New customer  
4  New customer  

[5 rows x 21 columns]

In [20]:
# You learned already that to access a single column in a dataframe, you can index it like so: df['column']

In [21]:
# check the accuracy of the 'loyalty_flag' column with output from columns 'user_id', 'max_order', and 'loyalty_flag' column

df_orders_products_merged[['user_id', 'max_order', 'loyalty_flag']].sample(30)

# The first set of brackets is the same as the brackets used in df['column'] above,
# while the second set of brackets is indicating that what you’re indexing is a list of multiple columns (and lists are always included within brackets).
# This is why there are two sets of brackets.

user_id  max_order      loyalty_flag
10117086    82650         42    Loyal customer
21789032   110131         30  Regular customer
18331742    83243         18  Regular customer
11855395   116255         98    Loyal customer
29623616    76011         17  Regular customer
28020391   180873          7      New customer
27935382   182873         96    Loyal customer
19758403    13084         18  Regular customer
10457404    55164         25  Regular customer
20778400    52331         36  Regular customer
3392860     65781         53    Loyal customer
18126508    77620         68    Loyal customer
29494698   189228         30  Regular customer
4960824     24650         23  Regular customer
9981484     38752         27  Regular customer
21002144   184566         38  Regular customer
2382658    123138         19  Regular customer
6057118     16189         11  Regular customer
3794461    165750         14  Regular customer
32325003   135079         52    Loyal customer
26241860    94330         99    Loyal customer
590593     144313          7      New customer
9325847     46750         26  Regular customer
22031185    65678         12  Regular customer
15382160   146193         58    Loyal customer
2121521    115781         62    Loyal customer
11794150   110833         52    Loyal customer
7457855     96252         41    Loyal customer
12031872    21208         13  Regular customer
26025176   197528         13  Regular customer

In [ ]:
# And that brings you to the end of the three-step process!
# 1. You started by grouping your data,
# 2. followed up by applying an aggregation function,
# 3. and finished everything off by using the resulting aggregation to create a flag column

# Task 4.8

# Step 1

Create a new notebook for this task. Be sure to import the relevant libraries, along with your ords_prods_merge dataframe, which should include your newly derived columns from the previous Exercise.

In [ ]:
# DONE

# Step 2

In this Exercise, you learned how to find the aggregated mean of the “order_number” column grouped by “department_id” for a subset of your dataframe. Now, repeat this process for the entire dataframe.

In [22]:
df_orders_products_merged.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

# Step 3

Analyze the result. How do the results for the entire dataframe differ from those of the subset? Include your comments in a markdown cell below the executed code.

In [23]:
# This is the subset one (I decided to run again instead of scrolling up)

df.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
4                18.825780
7                17.472355
13               17.993423
14               19.246334
16               19.463012
17               11.294069
19               19.305237
20               17.599636

Answer: The average result of order_number from the subset (1000000) is greater than the average result of entire dataframe (except the department_id: 17)

# Step 4

Follow the instructions in the Exercise for creating a loyalty flag for existing customers using the transform() and loc() functions.

In [ ]:
# DONE (Output [26])

# Step 5

Determine the difference between the spending habits of the three types of customers you identified.

Determine is whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers.

In [24]:
print('The mean, min and max of prices grouped by the loyalty_flag column in df_orders_products_merged:')
df_orders_products_merged.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max']})

The mean, min and max of prices grouped by the loyalty_flag column in df_orders_products_merged:


prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.386336  1.0  99999.0
New customer      13.294670  1.0  99999.0
Regular customer  12.495717  1.0  99999.0

Observation:

1. It is interesting to see that new customers have a higher average purchase price than the other 2 customer categories and even loyal customers have a lower average purchase than the others

2. The maximum price here seems odd compared to the mean value itself. The value of 99.999 has skewed the distribution so further analysis for this value will be performed.

In [25]:
# Run descriptive statistics on the 'prices' column

print('Descriptive statistics of the prices column:')
df_orders_products_merged['prices'].describe()

Descriptive statistics of the prices column:


count    3.240486e+07
mean     1.198023e+01
std      4.956554e+02
min      1.000000e+00
25%      4.200000e+00
50%      7.400000e+00
75%      1.130000e+01
max      9.999900e+04
Name: prices, dtype: float64

Observation:

1. The average (mean) price is 11.9 and the maximum price is 99.999, through this we can see that there is a huge difference in this data

2. The standard deviation has been confirmed which has a std of 495.655 indicating that we have a lot of variation in this data (especially large values that are higher than the mean)

In [26]:
# Check for abnormal values in the 'prices' column

print('Check for abnormal values in the prices column:')
df_orders_products_merged['prices'].sort_values(ascending = False)

Check for abnormal values in the prices column:


29165979    99999.0
29165985    99999.0
29165519    99999.0
29165518    99999.0
29165517    99999.0
             ...   
7870751         1.0
7870750         1.0
7870749         1.0
7870748         1.0
30034999        1.0
Name: prices, Length: 32404859, dtype: float64

In [27]:
print('Check for abnormal values in the prices column:')

df_orders_products_merged['prices'].drop_duplicates().sort_values(ascending = False)

Check for abnormal values in the prices column:


29165516    99999.0
10030345    14900.0
18504754       25.0
25758883       24.9
28425432       24.8
             ...   
3384024         1.4
632766          1.3
5488887         1.2
5147325         1.1
60725           1.0
Name: prices, Length: 242, dtype: float64

Observation: The 99.999 and 14.900 prices look weird (that's a lot higher than the others)

In [28]:
# Get the output of df_orders_products_merged when the value in the 'prices' column is 99999 or 14900

max_price = df_orders_products_merged.loc[df_orders_products_merged['prices'].isin([99999, 14900])]

In [33]:
# Sample output:

print('Sample output of df_orders_products_merged when the value in the prices column is either 99999 or 14900:')
max_price.sample(5)

Sample output of df_orders_products_merged when the value in the prices column is either 99999 or 14900:


order_id  user_id eval_set  order_number  orders_day_of_week  \
10032015   1079810    82185    prior            25                   1   
10032528   2845776   102256    prior            75                   2   
29165766    167037    92658    prior            27                   5   
10032981   1232320   123336    prior             2                   3   
10032854   2002312   114571    prior             3                   4   

          order_hour_of_day  days_since_last_order  product_id  \
10032015                 11                    7.0       21553   
10032528                 11                    5.0       21553   
29165766                  7                    6.0       33664   
10032981                  8                    4.0       21553   
10032854                  7                   14.0       21553   

          add_to_cart_order  reordered  ...                      product_name  \
10032015                  2          1  ...  Lowfat 2% Milkfat Cottage Cheese   
10032528                  1          1  ...  Lowfat 2% Milkfat Cottage Cheese   
29165766                  2          1  ...             2 % Reduced Fat  Milk   
10032981                  1          0  ...  Lowfat 2% Milkfat Cottage Cheese   
10032854                 15          1  ...  Lowfat 2% Milkfat Cottage Cheese   

         aisle_id  department_id   prices  _merge     busiest_day  \
10032015      108             16  14900.0    both  Regularly busy   
10032528      108             16  14900.0    both  Regularly busy   
29165766       84             16  99999.0    both  Regularly busy   
10032981      108             16  14900.0    both  Regularly busy   
10032854      108             16  14900.0    both      Least busy   

             busiest_days busiest_period_of_day max_order      loyalty_flag  
10032015     Busiest days           Most orders        41    Loyal customer  
10032528   Regularly busy           Most orders        99    Loyal customer  
29165766   Regularly busy        Average orders        42    Loyal customer  
10032981  Least busy days        Average orders        19  Regular customer  
10032854  Least busy days        Average orders         8      New customer  

[5 rows x 21 columns]

In [29]:
# Descriptive statistics on max_price

print('Descriptive statistics on max_price:')
max_price.describe()

# I need to scroll to the right to find the price, department_id and product_id

Descriptive statistics on max_price:


order_id        user_id  order_number  orders_day_of_week  \
count  5.127000e+03    5127.000000   5127.000000         5127.000000   
mean   1.708876e+06  106117.185684     15.302126            2.750536   
std    9.838956e+05   58594.982659     16.691849            2.010892   
min    4.300000e+01      17.000000      1.000000            0.000000   
25%    8.642320e+05   58136.000000      4.000000            1.000000   
50%    1.693799e+06  106021.000000      9.000000            3.000000   
75%    2.567244e+06  156389.000000     20.000000            5.000000   
max    3.421047e+06  206049.000000     99.000000            6.000000   

       order_hour_of_day  days_since_last_order    product_id  \
count        5127.000000            4776.000000   5127.000000   
mean           13.473376              11.899079  23201.815682   
std             4.137326               9.068019   4153.742747   
min             0.000000               0.000000  21553.000000   
25%            10.000000               5.000000  21553.000000   
50%            13.000000               8.000000  21553.000000   
75%            16.000000              16.000000  21553.000000   
max            23.000000              30.000000  33664.000000   

       add_to_cart_order    reordered  Unnamed: 0_y     aisle_id  \
count        5127.000000  5127.000000   5127.000000  5127.000000   
mean            7.986347     0.613614  23202.951824   104.732592   
std             6.868889     0.486968   4154.085720     8.231346   
min             1.000000     0.000000  21554.000000    84.000000   
25%             3.000000     0.000000  21554.000000   108.000000   
50%             6.000000     1.000000  21554.000000   108.000000   
75%            11.000000     1.000000  21554.000000   108.000000   
max            72.000000     1.000000  33666.000000   108.000000   

       department_id        prices    max_order  
count         5127.0   5127.000000  5127.000000  
mean            16.0  26485.547494    29.791691  
std              0.0  29186.636448    24.929207  
min             16.0  14900.000000     3.000000  
25%             16.0  14900.000000    12.000000  
50%             16.0  14900.000000    21.000000  
75%             16.0  14900.000000    41.000000  
max             16.0  99999.000000    99.000000

Observation: Incredibly high prices (14.900 and 99.999) both from the department_id 16 with product_id 21553 and 33664

In [35]:
# Find the product_name with product_id 21553 and 33664

df_orders_products_merged.loc[df_orders_products_merged['product_id'].isin([21553, 33664])]

order_id  user_id eval_set  order_number  orders_day_of_week  \
10030345    912404       17    prior            12                   2   
10030346    603376       17    prior            22                   6   
10030347   3264360      135    prior             2                   2   
10030348    892534      135    prior             3                   0   
10030349    229704      342    prior             8                   1   
...            ...      ...      ...           ...                 ...   
29166209   2249946   204099    prior            29                   0   
29166210   2363282   204099    prior            31                   0   
29166211   3181945   204395    prior            13                   3   
29166212   2486215   205227    prior             7                   3   
29166213   2232716   205227    prior            10                   0   

          order_hour_of_day  days_since_last_order  product_id  \
10030345                 14                    5.0       21553   
10030346                 16                    4.0       21553   
10030347                 21                   13.0       21553   
10030348                  8                   12.0       21553   
10030349                 19                   30.0       21553   
...                     ...                    ...         ...   
29166209                  8                    4.0       33664   
29166210                  9                    2.0       33664   
29166211                 15                    8.0       33664   
29166212                 20                    4.0       33664   
29166213                 11                    4.0       33664   

          add_to_cart_order  reordered  ...                      product_name  \
10030345                  5          0  ...  Lowfat 2% Milkfat Cottage Cheese   
10030346                  3          1  ...  Lowfat 2% Milkfat Cottage Cheese   
10030347                  6          0  ...  Lowfat 2% Milkfat Cottage Cheese   
10030348                  3          1  ...  Lowfat 2% Milkfat Cottage Cheese   
10030349                  9          0  ...  Lowfat 2% Milkfat Cottage Cheese   
...                     ...        ...  ...                               ...   
29166209                  1          0  ...             2 % Reduced Fat  Milk   
29166210                  1          1  ...             2 % Reduced Fat  Milk   
29166211                 25          0  ...             2 % Reduced Fat  Milk   
29166212                  8          0  ...             2 % Reduced Fat  Milk   
29166213                 23          1  ...             2 % Reduced Fat  Milk   

         aisle_id  department_id   prices  _merge     busiest_day  \
10030345      108             16  14900.0    both  Regularly busy   
10030346      108             16  14900.0    both  Regularly busy   
10030347      108             16  14900.0    both  Regularly busy   
10030348      108             16  14900.0    both     Busiest day   
10030349      108             16  14900.0    both  Regularly busy   
...           ...            ...      ...     ...             ...   
29166209       84             16  99999.0    both     Busiest day   
29166210       84             16  99999.0    both     Busiest day   
29166211       84             16  99999.0    both  Regularly busy   
29166212       84             16  99999.0    both  Regularly busy   
29166213       84             16  99999.0    both     Busiest day   

             busiest_days busiest_period_of_day max_order      loyalty_flag  
10030345   Regularly busy           Most orders        40  Regular customer  
10030346   Regularly busy           Most orders        40  Regular customer  
10030347   Regularly busy        Average orders         4      New customer  
10030348     Busiest days        Average orders         4      New customer  
10030349     Busiest days        Average orders        16  Regular customer  
...                   ...                   ...       ... 

Observation:

1. The price of 14.900 comes from Lowfat 2% Milkfat Cottage Cheese (with product_id 21553)

2. The price of 99.999 comes from 2 % Reduced Fat Milk (with product_id 33664)

In [30]:
# Create a filter 'dairy' for records with the department_id value of 16, 
# and then 'dairy_without_weird' to exclude records (without very high prices)
# using ~ to exclude the data

dairy = df_orders_products_merged.loc[df_orders_products_merged['department_id'].isin([16])]
dairy_without_weird = dairy.loc[~dairy['prices'].isin([99999, 14900])]
print('Filter for records with the department_id value of 16, while excluding records without the unusually high prices:')
dairy_without_weird.sample(5)

Filter for records with the department_id value of 16, while excluding records without the unusually high prices:


order_id  user_id eval_set  order_number  orders_day_of_week  \
20312390    520508   158702    prior            24                   4   
5436513     195109    53928    prior             9                   0   
16733395   1401244     6786    prior             4                   0   
12387746   3412949    13628    prior            68                   1   
21908701    853832   162002    prior             1                   1   

          order_hour_of_day  days_since_last_order  product_id  \
20312390                 11                   25.0       38772   
5436513                  14                   10.0       10895   
16733395                 12                   22.0       36011   
12387746                 11                    3.0       23644   
21908701                 12                    NaN       13237   

          add_to_cart_order  reordered  ...  \
20312390                  2          1  ...   
5436513                  15          1  ...   
16733395                 16          0  ...   
12387746                 15          0  ...   
21908701                  2          0  ...   

                               product_name aisle_id  department_id  prices  \
20312390                          Provolone       21             16     4.3   
5436513      Mexican Finely Shredded Cheese       21             16     6.6   
16733395              Organic Fat Free Milk       84             16     3.2   
12387746  Sharp Cheddar Thick Slices Cheese       21             16     2.4   
21908701            Original Egg Substitute       86             16     1.7   

          _merge     busiest_day     busiest_days busiest_period_of_day  \
20312390    both      Least busy  Least busy days           Most orders   
5436513     both     Busiest day     Busiest days           Most orders   
16733395    both     Busiest day     Busiest days           Most orders   
12387746    both  Regularly busy     Busiest days           Most orders   
21908701    both  Regularly busy     Busiest days           Most orders   

         max_order      loyalty_flag  
20312390        28  Regular customer  
5436513         97    Loyal customer  
16733395        16  Regular customer  
12387746        69    Loyal customer  
21908701        25  Regular customer  

[5 rows x 21 columns]

In [37]:
# Check to see if the filter is working properly

print('Maximum price in the dairy_without_weird filter:')
dairy_without_weird['prices'].max()


Maximum price in the dairy_without_weird filter:


15.0

Observation: After excluding unusual values, the max price becomes more reasonable

In [31]:
# First: Observation on 2 % Reduced Fat Milk (with product_id 33664)
# Create a filter for products that contain 'Reduced Fat', '2' and 'Milk' in the name, in dairy_without_weird

reduced_fat_milk = dairy_without_weird.query('product_name.str.contains("Reduced Fat") & product_name.str.contains("2") & product_name.str.contains("Milk")')


In [39]:
print('Sample output from reduced_fat_milk:')
reduced_fat_milk.sample(10)

Sample output from reduced_fat_milk:


order_id  user_id eval_set  order_number  orders_day_of_week  \
11702323    499959   111119    prior            26                   1   
22331772   3417063    89079    prior            31                   2   
16346406    875671   148471    prior             7                   4   
12866036    935039   112663    prior            34                   0   
20167323   1494690   110266    prior             6                   2   
20791733   1384196   129379    prior             7                   4   
14563025   1394986   162747    prior             4                   4   
15996872   2685478     5667    prior             1                   5   
16033765   1674804   162867    prior            21                   1   
16335201   2695166     9657    prior            12                   1   

          order_hour_of_day  days_since_last_order  product_id  \
11702323                 13                    7.0       32478   
22331772                  9                    6.0       28593   
16346406                 14                    5.0       40174   
12866036                 15                    6.0        1940   
20167323                 16                    7.0       13166   
20791733                 19                   18.0       19820   
14563025                  9                    7.0       23909   
15996872                 12                    NaN        5785   
16033765                  8                   13.0        5785   
16335201                  6                    6.0       40174   

          add_to_cart_order  reordered  ...  \
11702323                  3          1  ...   
22331772                  3          1  ...   
16346406                  1          0  ...   
12866036                  3          1  ...   
20167323                  2          1  ...   
20791733                  4          0  ...   
14563025                  1          1  ...   
15996872                  1          0  ...   
16033765                  1          1  ...   
16335201                 15          1  ...   

                                  product_name aisle_id  department_id  \
11702323                   Reduced Fat 2% Milk       84             16   
22331772         Reduced Fat 2% Chocolate Milk       84             16   
16346406           2% Reduced Fat Organic Milk       84             16   
12866036           Organic 2% Reduced Fat Milk       84             16   
20167323  Organic Milk Reduced Fat, 2% Milkfat       84             16   
20791733         Grassmilk 2% Reduced Fat Milk       84             16   
14563025                   2% Reduced Fat Milk       84             16   
15996872           Organic Reduced Fat 2% Milk       84             16   
16033765           Organic Reduced Fat 2% Milk       84             16   
16335201           2% Reduced Fat Organic Milk       84             16   

          prices  _merge     busiest_day     busiest_days  \
11702323    13.9    both  Regularly busy     Busiest days   
22331772     1.7    both  Regularly busy   Regularly busy   
16346406    11.4    both      Least busy  Least busy days   
12866036     9.1    both     Busiest day     Busiest days   
20167323    12.8    both  Regularly busy   Regularly busy   
20791733    12.6    both      Least busy  Least busy days   
14563025     9.2    both      Least busy  Least busy days   
15996872     7.9    both  Regularly busy   Regularly busy   
16033765     7.9    both  Regularly busy     Busiest days   
16335201    11.4    both  Regularly busy     Busiest days   

         busiest_period_of_day max_order      loyalty_flag  
11702323           Most orders        39  Regular customer  
22331772           Most orders        68    Loyal customer  
16346406           Most orders        15  Regular customer  
12866036           Most orders        46    Loyal customer  
20167323           Most orders         6      New customer  
20791733        Average orders        18  Regular customer  
14563025           Most orders      

In [33]:
# Quick statistics in reduced_fat_milk

print('Quick statistics in reduced_fat_milk:')
reduced_fat_milk.describe()

Quick statistics in reduced_fat_milk:


order_id        user_id   order_number  orders_day_of_week  \
count  2.022380e+05  202238.000000  202238.000000       202238.000000   
mean   1.711414e+06  103040.384151      18.528684            2.748702   
std    9.876774e+05   59620.791055      18.288088            2.022770   
min    5.000000e+00      11.000000       1.000000            0.000000   
25%    8.586890e+05   51496.000000       5.000000            1.000000   
50%    1.706996e+06  102925.000000      12.000000            3.000000   
75%    2.566178e+06  154743.000000      26.000000            5.000000   
max    3.421036e+06  206209.000000      99.000000            6.000000   

       order_hour_of_day  days_since_last_order     product_id  \
count      202238.000000          190556.000000  202238.000000   
mean           13.181835              10.538445   18768.761009   
std             4.193794               8.380235   13345.840890   
min             0.000000               0.000000    1940.000000   
25%            10.000000               5.000000    5785.000000   
50%            13.000000               7.000000   19820.000000   
75%            16.000000              14.000000   32478.000000   
max            23.000000              30.000000   49319.000000   

       add_to_cart_order      reordered   Unnamed: 0_y       aisle_id  \
count      202238.000000  202238.000000  202238.000000  202238.000000   
mean            5.423051       0.782692   18769.890520      83.716626   
std             5.397505       0.412415   13347.113649       6.511429   
min             1.000000       0.000000    1940.000000      21.000000   
25%             2.000000       1.000000    5785.000000      84.000000   
50%             4.000000       1.000000   19821.000000      84.000000   
75%             7.000000       1.000000   32480.000000      84.000000   
max           137.000000       1.000000   49323.000000      91.000000   

       department_id         prices      max_order  
count       202238.0  202238.000000  202238.000000  
mean            16.0       9.086900      35.834769  
std              0.0       3.064869      26.069071  
min             16.0       1.200000       3.000000  
25%             16.0       7.900000      14.000000  
50%             16.0       9.200000      30.000000  
75%             16.0      11.400000      51.000000  
max             16.0      14.700000      99.000000

Observation: Now the mean, min and max prices for products that contain 'Reduced Fat', '2' and 'Milk' in the name are: 9.08, 1.2, 14.7. Since this result makes more sense, the price of product ID 33664, named '2 % Reduced Fat Milk' will be reduced from 99.999 to 9.99

In [34]:
# Second: Observation on Lowfat 2% Milkfat Cottage Cheese (with product_id 21553)
# Create a filter for products that contain 'Cottage Cheese', '2' and 'Milkfat' in the name, in dairy_without_weird

cottage_cheese = dairy_without_weird.query('product_name.str.contains("Cottage Cheese") & product_name.str.contains("2") & product_name.str.contains("Milkfat")')

In [35]:
print('Sample output from cottage_cheese:')
cottage_cheese.sample(10)

Sample output from cottage_cheese:


order_id  user_id eval_set  order_number  orders_day_of_week  \
16191815   2347131   114280    prior            37                   0   
16191549    434304   100917    prior             1                   1   
16191025   2051281    55312    prior            11                   3   
16191184    593767    71834    prior            82                   5   
16191977   2404040   129521    prior             3                   2   
16192743   2383896   198113    prior            26                   6   
16192580    651840   184063    prior             4                   6   
16192414   3029659   170297    prior            26                   1   
29899052   2307026    97795    prior            20                   5   
25215388   1379284   201247    prior            10                   2   

          order_hour_of_day  days_since_last_order  product_id  \
16191815                 14                    7.0       27790   
16191549                 21                    NaN       27790   
16191025                 11                   25.0       27790   
16191184                  9                    7.0       27790   
16191977                 21                    5.0       27790   
16192743                 15                   14.0       27790   
16192580                 15                    9.0       27790   
16192414                 12                   10.0       27790   
29899052                 10                    6.0        6852   
25215388                 16                    9.0       16257   

          add_to_cart_order  reordered  ...  \
16191815                  4          1  ...   
16191549                  1          0  ...   
16191025                  1          1  ...   
16191184                  2          1  ...   
16191977                  1          0  ...   
16192743                 20          0  ...   
16192580                  5          1  ...   
16192414                  6          1  ...   
29899052                  3          1  ...   
25215388                  6          0  ...   

                                               product_name aisle_id  \
16191815        Small Curd Lowfat 2% Milkfat Cottage Cheese      108   
16191549        Small Curd Lowfat 2% Milkfat Cottage Cheese      108   
16191025        Small Curd Lowfat 2% Milkfat Cottage Cheese      108   
16191184        Small Curd Lowfat 2% Milkfat Cottage Cheese      108   
16191977        Small Curd Lowfat 2% Milkfat Cottage Cheese      108   
16192743        Small Curd Lowfat 2% Milkfat Cottage Cheese      108   
16192580        Small Curd Lowfat 2% Milkfat Cottage Cheese      108   
16192414        Small Curd Lowfat 2% Milkfat Cottage Cheese      108   
29899052         4% Milkfat Cottage Cheese Small Curd 22 OZ      108   
25215388  2% Milkfat Low Fat Large Curd Pot Style Cottag...      108   

          department_id  prices  _merge     busiest_day     busiest_days  \
16191815             16     7.7    both     Busiest day     Busiest days   
16191549             16     7.7    both  Regularly busy     Busiest days   
16191025             16     7.7    both  Regularly busy  Least busy days   
16191184             16     7.7    both  Regularly busy   Regularly busy   
16191977             16     7.7    both  Regularly busy   Regularly busy   
16192743             16     7.7    both  Regularly busy   Regularly busy   
16192580             16     7.7    both  Regularly busy   Regularly busy   
16192414             16     7.7    both  Regularly busy     Busiest days   
29899052             16     8.6    both  Regularly busy   Regularly busy   
25215388             16     5.3    both  Regularly busy   Regularly busy   

         busiest_period_of_day max_order      loyalty_flag  
16191815           Most orders        46    Loyal customer  
16191549        Average orders        15  Regular customer  
16191025           Most orders        20  Regular customer  
16191184           Most orders        85    Loyal customer  
16191977       

In [36]:
# Quick statistics in cottage_cheese

print('Quick statistics in cottage_cheese:')
cottage_cheese.describe()

Quick statistics in cottage_cheese:


order_id        user_id  order_number  orders_day_of_week  \
count  3.487000e+03    3487.000000   3487.000000         3487.000000   
mean   1.684106e+06  103778.688271     15.484084            2.849728   
std    9.785618e+05   58583.418992     16.808776            2.031540   
min    2.301000e+03      52.000000      1.000000            0.000000   
25%    8.438150e+05   55839.000000      4.000000            1.000000   
50%    1.673171e+06  107636.000000     10.000000            3.000000   
75%    2.535750e+06  151541.500000     20.000000            5.000000   
max    3.419042e+06  206072.000000     98.000000            6.000000   

       order_hour_of_day  days_since_last_order    product_id  \
count        3487.000000            3218.000000   3487.000000   
mean           13.257241              12.898073  22348.791225   
std             4.297401               9.231128   8826.332190   
min             0.000000               0.000000    195.000000   
25%            10.000000               6.000000  11203.000000   
50%            13.000000              10.000000  27790.000000   
75%            16.000000              19.000000  27790.000000   
max            23.000000              30.000000  31030.000000   

       add_to_cart_order    reordered  Unnamed: 0_y     aisle_id  \
count        3487.000000  3487.000000   3487.000000  3487.000000   
mean            5.840551     0.697448  22350.199312   107.650703   
std             5.955836     0.459429   8827.244775     5.502322   
min             1.000000     0.000000    194.000000    21.000000   
25%             2.000000     0.000000  11203.000000   108.000000   
50%             4.000000     1.000000  27792.000000   108.000000   
75%             8.000000     1.000000  27792.000000   108.000000   
max            57.000000     1.000000  31032.000000   108.000000   

       department_id       prices    max_order  
count         3487.0  3487.000000  3487.000000  
mean            16.0     7.228219    30.531116  
std              0.0     1.757900    25.413902  
min             16.0     2.400000     3.000000  
25%             16.0     7.700000    11.000000  
50%             16.0     7.700000    22.000000  
75%             16.0     7.700000    43.000000  
max             16.0    11.800000    99.000000

Observation: Now the mean, min and max prices for products that contain 'Cottage Cheese', '2' and 'Milkfat' in the name are: 7.23, 2.4, 11.8. Since this result makes more sense, the price of product ID 21553, named 'Lowfat 2% Milkfat Cottage Cheese' will be replaced from 14.900 and inputed by the mean value mentioned above

In [37]:
# Replace the price '99999' with '9.99' and impute the value '14900' with the mean price from cottage_cheese

df_orders_products_merged['prices'] = df_orders_products_merged['prices'].replace({99999: 9.99, 14900: cottage_cheese['prices'].mean()})

In [38]:
# Check the statistics of the new 'prices' column in df_orders_products_merged

print('The statistics of the new prices column in df_orders_products_merged:')
df_orders_products_merged['prices'].describe()

The statistics of the new prices column in df_orders_products_merged:


count    3.240486e+07
mean     7.790965e+00
std      4.241491e+00
min      1.000000e+00
25%      4.200000e+00
50%      7.400000e+00
75%      1.130000e+01
max      2.500000e+01
Name: prices, dtype: float64

Observation: Now the mean, std, min and max results make more sense than Output [30]

Step 5 (Updated: After this step --> replace({99999: 9.99, 14900: cottage_cheese['prices'].mean())

Determine the difference between the spending habits of the three types of customers you identified.

Determine is whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers.

In [39]:
# Basic statistics of df_orders_products_merged by 'loyalty_flag' with the updated 'prices' column

print('The mean, min and max of prices grouped by loyalty_flag with the updated prices column in df_orders_products_merged:')
df_orders_products_merged.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max']})

The mean, min and max of prices grouped by loyalty_flag with the updated prices column in df_orders_products_merged:


prices           
                      mean  min   max
loyalty_flag                         
Loyal customer    7.773526  1.0  25.0
New customer      7.801190  1.0  25.0
Regular customer  7.798239  1.0  25.0

Observation: Although the result remains the same (compared to Output [29]): new customer holds the highest purchase price followed by regular customer then loyal customer, now the mean value becomes lower and max value is more reasonable (instead of 99999, now 25)

# Step 6

Create a spending flag for each user based on the average price across all their orders using the following criteria:

1. If the mean of the prices of products purchased by a user is lower than 10, then flag them as a “Low spender.”
2. If the mean of the prices of products purchased by a user is higher than or equal to 10, then flag them as a “High spender.”

In [40]:
# Creating “average_price” column using transform()

df_orders_products_merged['average_price'] = df_orders_products_merged.groupby(['user_id'])['prices'].transform(np.mean)

In [41]:
df_orders_products_merged.head()

order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329        1    prior             1                   2   
1   2398795        1    prior             2                   3   
2    473747        1    prior             3                   3   
3   2254736        1    prior             4                   4   
4    431534        1    prior             5                   4   

   order_hour_of_day  days_since_last_order  product_id  add_to_cart_order  \
0                  8                    NaN         196                  1   
1                  7                   15.0         196                  1   
2                 12                   21.0         196                  1   
3                  7                   29.0         196                  1   
4                 15                   28.0         196                  1   

   reordered  ...  aisle_id department_id  prices  _merge     busiest_day  \
0          0  ...        77             7     9.0    both  Regularly busy   
1          1  ...        77             7     9.0    both  Regularly busy   
2          1  ...        77             7     9.0    both  Regularly busy   
3          1  ...        77             7     9.0    both      Least busy   
4          1  ...        77             7     9.0    both      Least busy   

      busiest_days busiest_period_of_day max_order  loyalty_flag  \
0   Regularly busy        Average orders        10  New customer   
1  Least busy days        Average orders        10  New customer   
2  Least busy days           Most orders        10  New customer   
3  Least busy days        Average orders        10  New customer   
4  Least busy days           Most orders        10  New customer   

   average_price  
0       6.367797  
1       6.367797  
2       6.367797  
3       6.367797  
4       6.367797  

[5 rows x 22 columns]

In [49]:
# Creating a 'spending_flag' flag

df_orders_products_merged.loc[df_orders_products_merged['average_price'] < 10, 'spending_flag'] = 'Low spender'

In [42]:
# Creating a 'spending_flag' flag

df_orders_products_merged.loc[df_orders_products_merged['average_price'] >= 10, 'spending_flag'] = 'High spender'

In [51]:
df_orders_products_merged.head()

order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329        1    prior             1                   2   
1   2398795        1    prior             2                   3   
2    473747        1    prior             3                   3   
3   2254736        1    prior             4                   4   
4    431534        1    prior             5                   4   

   order_hour_of_day  days_since_last_order  product_id  add_to_cart_order  \
0                  8                    NaN         196                  1   
1                  7                   15.0         196                  1   
2                 12                   21.0         196                  1   
3                  7                   29.0         196                  1   
4                 15                   28.0         196                  1   

   reordered  ...  department_id prices  _merge     busiest_day  \
0          0  ...              7    9.0    both  Regularly busy   
1          1  ...              7    9.0    both  Regularly busy   
2          1  ...              7    9.0    both  Regularly busy   
3          1  ...              7    9.0    both      Least busy   
4          1  ...              7    9.0    both      Least busy   

      busiest_days busiest_period_of_day max_order  loyalty_flag  \
0   Regularly busy        Average orders        10  New customer   
1  Least busy days        Average orders        10  New customer   
2  Least busy days           Most orders        10  New customer   
3  Least busy days        Average orders        10  New customer   
4  Least busy days           Most orders        10  New customer   

  average_price  spending_flag  
0      6.367797    Low spender  
1      6.367797    Low spender  
2      6.367797    Low spender  
3      6.367797    Low spender  
4      6.367797    Low spender  

[5 rows x 23 columns]

In [43]:
# Check the frequency

df_orders_products_merged['spending_flag'].value_counts(dropna = False)

NaN             32285150
High spender      119709
Name: spending_flag, dtype: int64

# Step 7

Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column. The criteria for the flag should be as follows:

1. If the median of “days_since_last_order” is higher than 20, then the customer should be labeled a “Non-frequent customer.”
2. If the median is higher than 10 and lower than or equal to 20, then the customer should be labeled a “Regular customer.”
3. If the median is lower than or equal to 10, then the customer should be labeled a “Frequent customer.”

In [44]:
# Creating “median_prior_orders” column using transform()

df_orders_products_merged['median_prior_orders'] = df_orders_products_merged.groupby(['user_id'])['days_since_last_order'].transform(np.median)

In [45]:
df_orders_products_merged.head()

order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329        1    prior             1                   2   
1   2398795        1    prior             2                   3   
2    473747        1    prior             3                   3   
3   2254736        1    prior             4                   4   
4    431534        1    prior             5                   4   

   order_hour_of_day  days_since_last_order  product_id  add_to_cart_order  \
0                  8                    NaN         196                  1   
1                  7                   15.0         196                  1   
2                 12                   21.0         196                  1   
3                  7                   29.0         196                  1   
4                 15                   28.0         196                  1   

   reordered  ...  prices _merge     busiest_day     busiest_days  \
0          0  ...     9.0   both  Regularly busy   Regularly busy   
1          1  ...     9.0   both  Regularly busy  Least busy days   
2          1  ...     9.0   both  Regularly busy  Least busy days   
3          1  ...     9.0   both      Least busy  Least busy days   
4          1  ...     9.0   both      Least busy  Least busy days   

   busiest_period_of_day max_order  loyalty_flag average_price spending_flag  \
0         Average orders        10  New customer      6.367797           NaN   
1         Average orders        10  New customer      6.367797           NaN   
2            Most orders        10  New customer      6.367797           NaN   
3         Average orders        10  New customer      6.367797           NaN   
4            Most orders        10  New customer      6.367797           NaN   

   median_prior_orders  
0                 20.5  
1                 20.5  
2                 20.5  
3                 20.5  
4                 20.5  

[5 rows x 24 columns]

In [46]:

# Creating a 'order_frequency_flag' flag

df_orders_products_merged.loc[df_orders_products_merged['median_prior_orders'] > 20, 'order_frequency_flag'] = 'Non-frequent customer'

In [47]:
# Creating a 'order_frequency_flag' flag

df_orders_products_merged.loc[(df_orders_products_merged['median_prior_orders'] > 10) & (df_orders_products_merged['median_prior_orders'] <= 20), 'order_frequency_flag'] = 'Regular customer'

In [48]:
# Creating a 'order_frequency_flag' flag

df_orders_products_merged.loc[df_orders_products_merged['median_prior_orders'] <= 10, 'order_frequency_flag'] = 'Frequent customer'

In [49]:
df_orders_products_merged.head()

order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329        1    prior             1                   2   
1   2398795        1    prior             2                   3   
2    473747        1    prior             3                   3   
3   2254736        1    prior             4                   4   
4    431534        1    prior             5                   4   

   order_hour_of_day  days_since_last_order  product_id  add_to_cart_order  \
0                  8                    NaN         196                  1   
1                  7                   15.0         196                  1   
2                 12                   21.0         196                  1   
3                  7                   29.0         196                  1   
4                 15                   28.0         196                  1   

   reordered  ...  _merge     busiest_day     busiest_days  \
0          0  ...    both  Regularly busy   Regularly busy   
1          1  ...    both  Regularly busy  Least busy days   
2          1  ...    both  Regularly busy  Least busy days   
3          1  ...    both      Least busy  Least busy days   
4          1  ...    both      Least busy  Least busy days   

   busiest_period_of_day  max_order  loyalty_flag average_price spending_flag  \
0         Average orders         10  New customer      6.367797           NaN   
1         Average orders         10  New customer      6.367797           NaN   
2            Most orders         10  New customer      6.367797           NaN   
3         Average orders         10  New customer      6.367797           NaN   
4            Most orders         10  New customer      6.367797           NaN   

  median_prior_orders   order_frequency_flag  
0                20.5  Non-frequent customer  
1                20.5  Non-frequent customer  
2                20.5  Non-frequent customer  
3                20.5  Non-frequent customer  
4                20.5  Non-frequent customer  

[5 rows x 25 columns]

In [50]:
# Check the frequency

df_orders_products_merged['order_frequency_flag'].value_counts(dropna = False)

Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
NaN                             5
Name: order_frequency_flag, dtype: int64

In [51]:
# Check the null values

print('Records in df_orders_products_merged with null values in column order_frequency_flag:')
df_orders_products_merged[df_orders_products_merged['order_frequency_flag'].isnull()][['user_id', 'days_since_last_order', 'median_prior_orders', 'order_frequency_flag']]

Records in df_orders_products_merged with null values in column order_frequency_flag:


user_id  days_since_last_order  median_prior_orders  \
13645692   159838                    NaN                  NaN   
17251990   159838                    NaN                  NaN   
17622767   159838                    NaN                  NaN   
24138593   159838                    NaN                  NaN   
25880002   159838                    NaN                  NaN   

         order_frequency_flag  
13645692                  NaN  
17251990                  NaN  
17622767                  NaN  
24138593                  NaN  
25880002                  NaN

In [52]:
# Check user_id 159838

df_orders_products_merged[df_orders_products_merged['user_id'] == 159838]

order_id  user_id eval_set  order_number  orders_day_of_week  \
13645692    895835   159838    prior             1                   0   
17251990    895835   159838    prior             1                   0   
17622767    895835   159838    prior             1                   0   
24138593    895835   159838    prior             1                   0   
25880002    895835   159838    prior             1                   0   

          order_hour_of_day  days_since_last_order  product_id  \
13645692                 17                    NaN       10749   
17251990                 17                    NaN       33401   
17622767                 17                    NaN       23695   
24138593                 17                    NaN       21334   
25880002                 17                    NaN       22198   

          add_to_cart_order  reordered  ...  _merge  busiest_day  \
13645692                  3          0  ...    both  Busiest day   
17251990                  6          0  ...    both  Busiest day   
17622767                  2          0  ...    both  Busiest day   
24138593                  5          0  ...    both  Busiest day   
25880002                  1          0  ...    both  Busiest day   

          busiest_days  busiest_period_of_day  max_order  loyalty_flag  \
13645692  Busiest days         Average orders          1  New customer   
17251990  Busiest days         Average orders          1  New customer   
17622767  Busiest days         Average orders          1  New customer   
24138593  Busiest days         Average orders          1  New customer   
25880002  Busiest days         Average orders          1  New customer   

         average_price spending_flag median_prior_orders  order_frequency_flag  
13645692          7.42           NaN                 NaN                   NaN  
17251990          7.42           NaN                 NaN                   NaN  
17622767          7.42           NaN                 NaN                   NaN  
24138593          7.42           NaN                 NaN                   NaN  
25880002          7.42           NaN                 NaN                   NaN  

[5 rows x 25 columns]

#### Observation: user_id 159838 seems to be a new customer. Therefore, I leave the value of NaN as it is

In [23]:
df_orders_products_merged.shape

(32404859, 21)

# Step 8

#### Ensure your notebook is clean and structured and that your code is well commented.

# Step 9

Export your dataframe as a pickle file and store it correctly in your “Prepared Data” folder.

In [53]:
df_orders_products_merged.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_merged_updated_2.pkl'))